dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/RpcStatus.java

```diff
@@ -111,7 +111,7 @@ public static void beginCount(URL url, String methodName) {
     public static boolean beginCount(URL url, String methodName, int max) {
         RpcStatus appStatus = getStatus(url);
         RpcStatus methodStatus = getStatus(url, methodName);
-        if (methodStatus.active.incrementAndGet() > max) {
+        if (max > 0 && methodStatus.active.incrementAndGet() > max) {

```

Comment:
If we add max>0 in the condition there it will cause the issue because in finally block we always reducing the endCount irrespective of max value. 

e.g. 
1)max value is 0 so in RpcStatus.java's begin count we will not increment the counter but in ActiveLimitFilter's invoke's finally block we will be always reducing so the final value for single invocation will landed up -1 active count.

I think not  including max>0 and adding 
} finally {      
    if (max > 0) {
                RpcStatus.endCount(url, methodName, System.currentTimeMillis() - begin, isSuccess);
                synchronized (count) {
                    count.notifyAll();
                   }
               }
         }
In ActiveLimitFilter should also work, but I beleive you can rewrite this into more clear way then we suggest.


I think we might have another issue too (not because of your changes), in a scenario if max>0 (in RpcStatus.java) get satisfied (or lets max>0 condition is not even present in code as earlier version) and the `methodStatus.active.incrementAndGet() > max` return `false` then code of else block will be executing and again it will increment the count. So for one call we are eventually incrementing counter twice. Do you feel this might be also a another problem in future?Correct me if my understanding not correct.

Reply:
first of all, `endCount` does more than limiting the traffic only, it is also responsible for running statistics, see below:

```java
    private static void endCount(RpcStatus status, long elapsed, boolean succeeded) {
        status.active.decrementAndGet();
        status.total.incrementAndGet();
        status.totalElapsed.addAndGet(elapsed);
        if (status.maxElapsed.get() < elapsed) {
            status.maxElapsed.set(elapsed);
        }
        if (succeeded) {
            if (status.succeededMaxElapsed.get() < elapsed) {
                status.succeededMaxElapsed.set(elapsed);
            }
        } else {
            status.failed.incrementAndGet();
            status.failedElapsed.addAndGet(elapsed);
            if (status.failedMaxElapsed.get() < elapsed) {
                status.failedMaxElapsed.set(elapsed);
            }
        }
    }
```

That's the reason we need to make sure `startCount` and `endCount` execute no matter `max` limit is set or not, but at the same time, we also need to make sure the filter will block incoming request if `max` limit exceeds.

Reply:
In my current implementation:

```java
if (max > 0 && methodStatus.active.incrementAndGet() > max) {
   return false;
} else {
  return true;
}
```

it satisfies what I explained above. When limit exceeds, it will return false, and since it is atomic operation, it can make sure further incoming requests to wait. If limit doesn't exceeds or is not set, then it will return true and allow the requests to continue, instead of wait.

More important, if we decide to allow the request to proceed, we need to make sure 'startCount' and 'endCount' to be called in pair.

Reply:
can I consider this topic is resolved?

Reply:
I see where the problem is. Pls. review the latest commit.